### PR TITLE
Use Context for Timeout Handling

### PR DESCRIPTION
**Pull request by kolanos**
This wires up `context` to HTTP requests made by the extension to cancel them when a context deadline is reached or the context is cancelled. The existing timeout logic could be greatly simplified using contexts. This will require further refactoring.

### CODE REVIEW

**Review by mrickard — DISMISSED** [review]

File: lambda/logserver/logserver.go [hunk 1]
```diff
@@ -149,7 +156,12 @@ func (ls *LogServer) handler(res http.ResponseWriter, req *http.Request) {
 }
 
 func Start() (*LogServer, error) {
-	return startInternal(defaultHost)
+	// This should only be used for testing purposes
+	host, hostOverride := os.LookupEnv("NEW_RELIC_LOG_SERVER_HOST_OVERRIDE")
```
Comment (MattWhelan) [inline lambda/logserver/logserver.go:160]: Can we push this into the configuration object, so that we don't have env vars everywhere?

**Comment by mrickard** [comment]
Would invoking .Panic interrupt the execution of the Lambda function?

**Comment by kolanos** [comment]
@mrickard The extension runs in a separate process from the lambda function. So the panics won't directly affect it. However, the `InitError` calls tell Lambda that something went wrong during initialization. After making this call, Lambda will restart the extension process anyway, so the panics are really just there to ensure execution is interrupted at that point.

File: main.go [hunk 2]
```diff
@@ -111,123 +151,134 @@ func main() {
...
-			}
-			// Tear down the timer
-			if !timeout.Stop() {
-				<-timeout.C
-			}
+			// Set the timeout timer for a smidge before the actual timeout;
+			// we can recover from early.
+			timeoutWatchBegins := time.Millisecond * 100
+			timeout := timeoutInstant.Sub(time.Now()) - timeoutWatchBegins
+			util.Debugf("Timeout in %v", timeout)
```
Comment (MattWhelan) [inline main.go:255]: That log line is going to be really chatty. Moreso that the debug level has been, historically. Let's not ship with this in place, if possible.

File: main.go [hunk 3]
```diff
@@ -111,123 +151,134 @@ func main() {
...
+				} else if event.ShutdownReason == api.Failure && lastRequestId != "" {
+					// Synthesize a generic platform error. Probably an OOM, though it could be any runtime crash.
+					errorMessage := fmt.Sprintf("RequestId: %s A platform error caused a shutdown", lastRequestId)
+					batch.AddTelemetry(lastRequestId, []byte(errorMessage))
+				}
+
+				pollLogServer(logServer, batch)
+				finalHarvest := batch.Close()
+				shipHarvest(ctx, finalHarvest, telemetryClient, invokedFunctionARN)
+				return eventCounter
```
Comment (MattWhelan) [inline main.go:238]: I think we need to shut down the log server before the final poll. That way, when the platform sends us logs too late, the platform will notice that we're not around to receive them. If we do the final poll first, then we drop the last logs.

**Review by mrickard — DISMISSED** [review]
LGTM, though there are those codecov complaints.

**Review by MattWhelan — CHANGES_REQUESTED** [review]
Mostly looks good, though the diff is really hard to read with all the codecov warnings and such. 

I do think that it's critical that we not change the shutdown ordering though.

File: lambda/extension/client/client.go [hunk 4]
```diff
@@ -110,30 +116,31 @@ func (ic *InvocationClient) getNextEventURL() string {
...
-// getInitErrorURL returns the Lambda Extension next event URL
+// getInitErrorURL returns the Lambda Extension initialization error URL
 func (ic *InvocationClient) getInitErrorURL() string {
 	return fmt.Sprintf("http://%s/%s/extension/init/error", ic.baseUrl, ic.version)
 }
 
-// getExitErrorURL returns the Lambda Extension next event URL
+// getExitErrorURL returns the Lambda exit error URL
 func (ic *InvocationClient) getExitErrorURL() string {
 	return fmt.Sprintf("http://%s/%s/extension/exit/error", ic.baseUrl, ic.version)
```
Comment (mrickard) [inline lambda/extension/client/client.go:126]: Does it matter if this simulated string is http or https?

**Review by mrickard — APPROVED** [review]

File: lambda/extension/client/client.go (hunk 4 above)
Comment (kolanos) [reply lambda/extension/client/client.go:130]: This is a part of the Lambda spec. The endpoint is local to the Lambda extension.

**Review by mrickard — APPROVED** [review]

**Pull request merged**